### ISSUE TITLE
swap_backgrounds not working

### ISSUE THREAD

**Issue by arkrost** [AlexvZyl/nordic.nvim#108]
I'm using lazy.nvim. I found out that `swap_background` option is not working. I tried to swap them with `on_palette` option and figured out that it is called twice and therefore not working. I believe it is a cause why `swap_backgrounds` is broken.

```
    {
      'AlexvZyl/nordic.nvim',
      event = 'VeryLazy',
      config = function()
        local counter = 0
        require('nordic').setup({
          on_palette = function(palette)
            counter = counter + 1
            print(counter, palette.gray0, palette.black1)
            local gray0 = palette.gray0
            palette.gray0 = palette.black1
            palette.black1 = gray0
            return palette
          end,
          -- swap_backgrounds = true,
        })
        vim.cmd('colorscheme nordic')
      end,
    },

```

**Comment by 5-pebbles** [comment]
Disclaimer: I have not checked if this is the case with your issue, but...

A lot of the options don't work if you run setup more than once (but some do).

`extend_palette` is now being called twice after #107, thereby calling `on_palette` twice.

Either you're right, and it's getting swapped twice, or it's being called with the default value of false and then not updating.

Again, I have not checked if I am right; this is just a guess...

The whole thing is a stateful mess courtesy of Lua, and I think it needs a major refactor.

- I would be willing to refactor it but not if I need to split it up into a few dozen PRs, which will take weeks to merge.

**Comment by AlexvZyl** [comment]
Yes, pretty sure I messed up or misunderstand how state works in Lua.

Sorry for being silent, and thanks for sticking around @5-pebbles 

I am going to sort it out this weekend.

**Comment by 5-pebbles** [comment]
I would be happy to sort it out, however I don't like merge conflicts, so I have been waiting on #100.

@arkrost is right the values are being switched [color/init.lua - 23]:
```lua
    -- Swap background
    if O.swap_backgrounds then
        local gray0 = C.gray0
        C.gray0 = C.black1
        C.black1 = gray0
    end
```

This will be accidentally fixed in #100.

**Comment by 5-pebbles** [comment]
So I just looked it up...

This returns a ref
```lua
require("module")
```
This returns a clone
```lua
require("module").value
```
In colors/init.lua you have:

```lua
local O = require('nordic.config').options
```
So after the first import that value is not being updated, which would also explain why the tests are running with the same config twice.

#101 is using this:

```lua
local options = require('nordic.config').options
```
So its getting a updated clone each time.

Example: https://replit.com/@5-pebbles/LuaImportsAreHard#main.lua

**Comment by AlexvZyl** [comment]
> So I just looked it up...
> 
> This returns a ref:
> 
> ```lua
> require("module")
> ```

Are you sure?  I thought all tables are references?

**Comment by AlexvZyl** [comment]
> I'm using lazy.nvim. I found out that `swap_background` option is not working. I tried to swap them with `on_palette` option and figured out that it is called twice and therefore not working. I believe it is a cause why `swap_backgrounds` is broken.
> 
> ```
>     {
>       'AlexvZyl/nordic.nvim',
>       event = 'VeryLazy',
>       config = function()
>         local counter = 0
>         require('nordic').setup({
>           on_palette = function(palette)
>             counter = counter + 1
>             print(counter, palette.gray0, palette.black1)
>             local gray0 = palette.gray0
>             palette.gray0 = palette.black1
>             palette.black1 = gray0
>             return palette
>           end,
>           -- swap_backgrounds = true,
>         })
>         vim.cmd('colorscheme nordic')
>       end,
>     },
> ```

Oh, yeah... That's not good.  Sorry just properly read this.  Though the problem was something else this entire time.

**Comment by AlexvZyl** [comment]
Sorry for the delay.  Should be fixed in the next merge.

**Comment by 5-pebbles** [comment]
> Are you sure? I thought all tables are references?

Yup your right, but what is it?

This import is definitely the changing factor, I tested that.

As far as I know only other thing that it could be, is changing the order of imports.

I don't even remember why I moved the import location but it must have caused me some issue.

I will open a issue for a refactor just to make the code more clear, so we can avoid issues like this in the future.
